lib/Runtime/Library/RegexHelper.cpp

```diff
@@ -792,8 +782,234 @@ namespace Js
+    }
+
+    template<typename ReplacementFn>
+    Var RegexHelper::RegexEs6ReplaceImpl(ScriptContext* scriptContext, RecyclableObject* thisObj, JavascriptString* input, ReplacementFn appendReplacement, bool noResult)
+    {
+        bool global = JavascriptRegExp::GetGlobalProperty(thisObj, scriptContext);
+        bool unicode = false; // Dummy value. It isn't used below unless "global" is "true".
+        if (global)
+        {
+            unicode = JavascriptRegExp::GetUnicodeProperty(thisObj, scriptContext);
+            JavascriptRegExp::SetLastIndexProperty(thisObj, TaggedInt::ToVarUnchecked(0), scriptContext);
+        }
+
+        JavascriptString* accumulatedResult = nullptr;
+
+        BEGIN_TEMP_ALLOCATOR(tempAlloc, scriptContext, L"RegexHelper")
+        {
+            JsUtil::List<RecyclableObject*, ArenaAllocator>* results =
+                JsUtil::List<RecyclableObject*, ArenaAllocator>::New(tempAlloc);
+
+            while (true)
+            {
+                PCWSTR varName = L"RegExp.prototype[Symbol.replace]";
+                Var result = JavascriptRegExp::CallExec(thisObj, input, varName, scriptContext);
+                if (JavascriptOperators::IsNull(result))
+                {
+                    break;
+                }
+
+                RecyclableObject* resultObj = ExecResultToRecyclableObject(result);
+
+                results->Add(resultObj);
+
+                if (!global)
+                {
+                    break;
+                }
+
+                JavascriptString* matchStr = GetMatchStrFromResult(resultObj, scriptContext);
+                AdvanceLastIndex(thisObj, input, matchStr, unicode, scriptContext);
+            }
+
+            CompoundString::Builder<64 * sizeof(void *) / sizeof(wchar_t)> accumulatedResultBuilder(scriptContext);
+            CharCount inputLength = input->GetLength();
+            CharCount nextSourcePosition = 0;
+
+            size_t previousNumberOfCapturesToKeep = 0;
+            Var* captures = nullptr;
+
+            results->Map([&](int i, RecyclableObject* resultObj) {
+                int64 length = JavascriptConversion::ToLength(
+                    JavascriptOperators::GetProperty(resultObj, PropertyIds::length, scriptContext),
+                    scriptContext);
+                uint64 numberOfCaptures = (uint64) max(length - 1, (int64) 0);
+
+                JavascriptString* matchStr = GetMatchStrFromResult(resultObj, scriptContext);
+
+                int64 index = JavascriptConversion::ToLength(
+                    JavascriptOperators::GetProperty(resultObj, PropertyIds::index, scriptContext),
+                    scriptContext);
+                CharCount position = max(
+                    min(JavascriptRegExp::GetIndexOrMax(index), inputLength),
+                    (CharCount) 0);
+
+                // Capture groups can be referenced using at most two digits.
+                uint64 maxNumberOfCaptures = 99;
```

Comment:
I made the variable `const`, but kept it where it currently is since, I think, it's better for it to be closer to where it is used.